File: app/src/main/java/com/centaury/cataloguemovie/ui/detail/DetailMovieActivity.java
package com.centaury.cataloguemovie.ui.detail;

import android.os.Build;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;

import com.centaury.cataloguemovie.BuildConfig;
import com.centaury.cataloguemovie.R;
import com.centaury.cataloguemovie.data.local.entity.MovieEntity;
import com.centaury.cataloguemovie.data.local.entity.TVShowEntity;
import com.centaury.cataloguemovie.utils.AppConstants;
import com.centaury.cataloguemovie.utils.GlideApp;

import java.text.DateFormat;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Locale;
import java.util.concurrent.ExecutionException;

import javax.inject.Inject;

import dagger.android.DispatchingAndroidInjector;

public class DetailMovieActivity extends AppCompatActivity {

    @Inject
    DispatchingAndroidInjector<Fragment> dispatchingAndroidInjector;

    @Inject
    ViewModelProvider.Factory factory;

    private DateFormat inputDate = new SimpleDateFormat("yyyy-MM-dd", Locale.getDefault());
    private DateFormat outputDate = new SimpleDateFormat("dd MMMM yyyy", Locale.getDefault());

    private DetailMovieResponse detailMovieResponse;
    private DetailTVShowResponse detailTVShowResponse;
    private DetailMovieViewModel detailMovieViewModel;
    private String imageBaseUrl = BuildConfig.IMAGE_URL;
    private String sizeImage = AppConstants.SIZE_IMAGE;
    private int movieId;
    private int tvshowId;
    private Boolean isFavorite = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_detail_movie);
        setSupportActionBar(toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setHomeAsUpIndicator(R.drawable.ic_arrow_back_white_24dp);
        getSupportActionBar().setDisplayShowTitleEnabled(false);

        Window window = getWindow();
        window.setFlags(WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS, WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS);

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            View view = getWindow().getDecorView();
            view.setSystemUiVisibility(view.getSystemUiVisibility() | View.SYSTEM_UI_FLAG_LIGHT_NAVIGATION_BAR);
        }

        detailMovieViewModel = obtainViewModel(this);
        String language = Locale.getDefault().toLanguageTag();

        mClData.setVisibility(View.GONE);
        mShimmerViewContainer.setVisibility(View.VISIBLE);
        mShimmerViewContainer.startShimmer();

        movieId = getIntent().getIntExtra(AppConstants.DETAIL_EXTRA_MOVIE, 0);
        if (movieId != 0) {
            detailMovieViewModel.setMovieId(String.valueOf(movieId));
            detailMovieViewModel.getDetailMovie(language).observe(this, detailMovieResponse -> {
                mShimmerViewContainer.stopShimmer();
                mShimmerViewContainer.setVisibility(View.GONE);
                mClData.setVisibility(View.VISIBLE);
                itemMovie(detailMovieResponse);
                this.detailMovieResponse = detailMovieResponse;
                stateFavoriteMovie(movieId);
            });

        } else {
            tvshowId = getIntent().getIntExtra(AppConstants.DETAIL_EXTRA_TV_SHOW, 0);

            detailMovieViewModel.setTvshowId(String.valueOf(tvshowId));
            detailMovieViewModel.getDetailTVShow(language).observe(this, detailTVShowResponse -> {
                mShimmerViewContainer.stopShimmer();
                mShimmerViewContainer.setVisibility(View.GONE);
                mClData.setVisibility(View.VISIBLE);
                itemTVShow(detailTVShowResponse);
                this.detailTVShowResponse = detailTVShowResponse;
                stateFavoriteTVShow(tvshowId);
            });
        }

        setFavorite();
        mBtnFavorite.setScale(2.481f);
    }

    private void itemMovie(DetailMovieResponse movie) {
        mTxtTitledetail.setText(movie.getTitle());
        mTxtRatemovie.setText(String.valueOf(movie.getVoteAverage()));
        float movieRating = (float) (movie.getVoteAverage() / 2);
        mRbRatingdetail.setRating(movieRating);

        if (movie.getGenres().size() == 0) {
            mTxtGenredetail.setText(getResources().getString(R.string.txt_no_genre));
        } else {
            getGenresString(movie.getGenres());
        }

        if (movie.getOverview() == null || movie.getOverview().equals("")) {
            mTxtDescdetail.setText(getResources().getString(R.string.txt_no_desc));
        } else {
            mTxtDescdetail.setText(movie.getOverview());
        }

        GlideApp.with(this)
                .load(imageBaseUrl + sizeImage + movie.getPosterPath())
                .apply(RequestOptions.placeholderOf(R.drawable.ic_loading).error(R.drawable.ic_error))
                .into(mIvImgdetail);
        if (movie.getBackdropPath() != null) {
            GlideApp.with(this)
                    .load(imageBaseUrl + sizeImage + movie.getBackdropPath())
                    .apply(RequestOptions.placeholderOf(R.drawable.ic_loading).error(R.drawable.ic_error))
                    .into(mIvCoverdetail);
        } else {
            GlideApp.with(this)
                    .load(imageBaseUrl + sizeImage + movie.getPosterPath())
                    .apply(RequestOptions.placeholderOf(R.drawable.ic_loading).error(R.drawable.ic_error))
                    .into(mIvCoverdetail);
        }

        try {
            Date date = inputDate.parse(movie.getReleaseDate());
            String releaseDate = null;
            if (date != null) {
                releaseDate = outputDate.format(date);
            }
            mTxtDatedetail.setText(releaseDate);
        } catch (ParseException e) {
            e.printStackTrace();
        }
    }

    private void itemTVShow(DetailTVShowResponse tvshow) {
        mTxtTitledetail.setText(tvshow.getName());
        mTxtRatemovie.setText(String.valueOf(tvshow.getVoteAverage()));
        float movieRating = (float) (tvshow.getVoteAverage() / 2);
        mRbRatingdetail.setRating(movieRating);

        if (tvshow.getGenres().size() == 0) {
            mTxtGenredetail.setText(getResources().getString(R.string.txt_no_genre));
        } else {
            getGenresString(tvshow.getGenres());
        }

        if (tvshow.getOverview() == null || tvshow.getOverview().equals("")) {
            mTxtDescdetail.setText(getResources().getString(R.string.txt_no_desc));
        } else {
            mTxtDescdetail.setText(tvshow.getOverview());
        }

        GlideApp.with(this)
                .load(imageBaseUrl + sizeImage + tvshow.getPosterPath())
                .apply(RequestOptions.placeholderOf(R.drawable.ic_loading).error(R.drawable.ic_error))
                .into(mIvImgdetail);
        if (tvshow.getBackdropPath() != null) {
            GlideApp.with(this)
                    .load(imageBaseUrl + sizeImage + tvshow.getBackdropPath())
                    .apply(RequestOptions.placeholderOf(R.drawable.ic_loading).error(R.drawable.ic_error))
                    .into(mIvCoverdetail);
        } else {
            GlideApp.with(this)
                    .load(imageBaseUrl + sizeImage + tvshow.getPosterPath())
                    .apply(RequestOptions.placeholderOf(R.drawable.ic_loading).error(R.drawable.ic_error))
                    .into(mIvCoverdetail);
        }

        try {
            Date date = inputDate.parse(tvshow.getFirstAirDate());
            String releaseDate = null;
            if (date != null) {
                releaseDate = outputDate.format(date);
            }
            mTxtDatedetail.setText(releaseDate);
        } catch (ParseException e) {
            e.printStackTrace();
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        mShimmerViewContainer.startShimmer();
    }

    @Override
    public void onPause() {
        mShimmerViewContainer.stopShimmer();
        super.onPause();
    }

    private void getGenresString(List<GenresItem> itemList) {
        List<String> genreMovies = new ArrayList<>();
        try {
            if (itemList.size() == 1) {
                for (GenresItem genresItem : itemList) {
                    genreMovies.add(genresItem.getName());
                }
                mTxtGenredetail.setText(TextUtils.join(", ", genreMovies));
            } else {
                List<GenresItem> genreMore = itemList.subList(0, 2);
                for (GenresItem genresItem : genreMore) {
                    genreMovies.add(genresItem.getName());
                }
                mTxtGenredetail.setText(TextUtils.join(", ", genreMovies));
            }
        } catch (IndexOutOfBoundsException | IllegalArgumentException e) {
            System.out.println("Exception thrown : " + e);
        }
    }

    private void addFavorite() {
        if (movieId != 0) {
            MovieEntity movieEntity = new MovieEntity(detailMovieResponse.getId(), mTxtTitledetail.getText().toString(), detailMovieResponse.getOriginalTitle(),
                    detailMovieResponse.getPosterPath(), detailMovieResponse.getBackdropPath(), mTxtGenredetail.getText().toString(),
                    detailMovieResponse.getReleaseDate(), mTxtDescdetail.getText().toString(), detailMovieResponse.getVoteAverage());
            detailMovieViewModel.insertFavoriteMovie(movieEntity);
        } else {
            TVShowEntity tvShowEntity = new TVShowEntity(detailTVShowResponse.getId(), mTxtTitledetail.getText().toString(), detailTVShowResponse.getOriginalName(),
                    detailTVShowResponse.getPosterPath(), detailTVShowResponse.getBackdropPath(), mTxtGenredetail.getText().toString(),
                    detailTVShowResponse.getFirstAirDate(), mTxtDescdetail.getText().toString(), detailTVShowResponse.getVoteAverage());
            detailMovieViewModel.insertFavoriteTVShow(tvShowEntity);
        }

        Toast.makeText(this, getString(R.string.txt_movie_add), Toast.LENGTH_SHORT).show();
    }

    private void removeFavorite(int id) {
        if (movieId != 0) {
            final MovieEntity movieEntity;
            try {
                movieEntity = detailMovieViewModel.getDetailFavMovie(id);
                detailMovieViewModel.deleteFavoriteMovie(movieEntity);

            } catch (ExecutionException | InterruptedException e) {
                e.printStackTrace();
            }
        } else {
            final TVShowEntity tvShowEntity;
            try {
                tvShowEntity = detailMovieViewModel.getDetailFavTVShow(id);
                detailMovieViewModel.deleteFavoriteTVShow(tvShowEntity);

            } catch (ExecutionException | InterruptedException e) {
                e.printStackTrace();
            }
        }

        Toast.makeText(this, getString(R.string.txt_movie_remove), Toast.LENGTH_SHORT).show();
    }

    private void stateFavoriteMovie(int id) {
        final MovieEntity movieEntity;
        try {
            movieEntity = detailMovieViewModel.getDetailFavMovie(id);
            if (movieEntity != null) {
                isFavorite = true;
            }
            setFavorite();
        } catch (ExecutionException | InterruptedException e) {
            e.printStackTrace();
        }
    }

    private void stateFavoriteTVShow(int id) {
        final TVShowEntity tvShowEntity;
        try {
            tvShowEntity = detailMovieViewModel.getDetailFavTVShow(id);
            if (tvShowEntity != null) {
                isFavorite = true;
            }
            setFavorite();
        } catch (ExecutionException | InterruptedException e) {
            e.printStackTrace();
        }
    }

    private void setFavorite() {
        if (isFavorite) {
            mBtnFavorite.setSpeed(1f);
            mBtnFavorite.setProgress(1f);
        } else {
            mBtnFavorite.setProgress(0f);
            mBtnFavorite.setSpeed(-2f);
        }
    }

    @OnClick(R.id.btn_favorite)
    public void onClick(View v) {
        if (v.getId() == R.id.btn_favorite) {
            if (isFavorite) {
                if (movieId != 0) {
                    removeFavorite(movieId);
                } else {
                    removeFavorite(tvshowId);
                }
            } else {
                addFavorite();
            }

            isFavorite = !isFavorite;

            setFavorite();
            mBtnFavorite.playAnimation();
        }
    }
}

File: app/src/main/java/com/centaury/cataloguemovie/ui/detail/DetailMovieActivity.kt
package com.centaury.cataloguemovie.ui.detail

import android.os.Build
import android.os.Bundle
import android.view.MenuItem
import android.view.WindowManager
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil.setContentView
import androidx.lifecycle.ViewModelProvider
import com.centaury.cataloguemovie.MovieCatalogueApp
import com.centaury.cataloguemovie.R
import com.centaury.cataloguemovie.databinding.ActivityDetailMovieBinding
import com.centaury.cataloguemovie.di.component.DaggerDetailComponent
import com.centaury.cataloguemovie.utils.*
import com.centaury.domain.detail.model.Detail
import com.centaury.domain.movies.model.MoviesDB
import com.centaury.domain.tvshow.model.TVShowsDB
import javax.inject.Inject

@Suppress("NULLABILITY_MISMATCH_BASED_ON_JAVA_ANNOTATIONS")
class DetailMovieActivity : AppCompatActivity() {

    @Inject
    lateinit var viewModelFactory: ViewModelProvider.Factory

    private val detailMovieViewModel: DetailMovieViewModel by viewModels {
        viewModelFactory
    }

    private lateinit var binding: ActivityDetailMovieBinding

    private var movie: MoviesDB? = null
    private var tvShow: TVShowsDB? = null
    private var detailFavorite: Detail? = null
    private var isFavorite: Boolean = false
    private var movieId: Int? = null
    private var tvShowId: Int? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        initInjector()
        super.onCreate(savedInstanceState)
        binding = setContentView(this, R.layout.activity_detail_movie)
        showSystemUI()

        window.setFlags(
            WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS,
            WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS
        )
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
            window.navigationBarDividerColor = getColorFromAttr(R.attr.colorPrimary)
        }

        setSupportActionBar(binding.toolbar)
        supportActionBar?.apply {
            setDisplayHomeAsUpEnabled(true)
            setHomeAsUpIndicator(R.drawable.ic_arrow)
            setDisplayShowTitleEnabled(false)
        }

        checkConnection(this)
        initClick()
        initView(binding)
    }

    private fun initView(binding: ActivityDetailMovieBinding) {
        initParam()
        initObserver(binding)

        setFavorite()
        binding.btnFavorite.scale = 2.481f
    }

    private fun initParam() {
        movieId = intent.getIntExtra(DETAIL_EXTRA_MOVIE, 0)
        if (movieId != 0) {
            movieId?.let {
                detailMovieViewModel.getFavoriteMovieByIdContract(it)
                detailMovieViewModel.getDetailMovieContract(it)
            }
        } else {
            tvShowId = intent.getIntExtra(DETAIL_EXTRA_TV_SHOW, 0)
            tvShowId?.let {
                detailMovieViewModel.getFavoriteTVShowByIdContract(it)
                detailMovieViewModel.getDetailTVShowContract(it)
            }
        }
    }

    private fun initObserver(binding: ActivityDetailMovieBinding) {
        detailMovieViewModel.state.observe(this, { state ->
            when (state) {
                is LoaderState.ShowLoading -> {
                    binding.shimmerViewContainer.startShimmer()
                    binding.hasDetailMovies = true
                }
                is LoaderState.HideLoading -> {
                    binding.shimmerViewContainer.stopShimmer()
                    binding.hasDetailMovies = false
                }
            }
        })

        detailMovieViewModel.resultMovie.observe(this, { resultMovie ->
            binding.ivImgDetail.transitionName = resultMovie.title
            supportPostponeEnterTransition()
            showDetail(resultMovie)
            detailFavorite = resultMovie
        })

        detailMovieViewModel.errorMovie.observe(this, { errorMovie ->
            timberE(errorMovie)
        })

        detailMovieViewModel.resultTVShow.observe(this, { resultTVShow ->
            binding.ivImgDetail.transitionName = resultTVShow.title
            supportPostponeEnterTransition()
            showDetail(resultTVShow)
            detailFavorite = resultTVShow
        })

        detailMovieViewModel.errorTVShow.observe(this, { errorTVShow ->
            timberE(errorTVShow)
        })

        detailMovieViewModel.resultMovieById.observe(this, { movieById ->
            movie = movieById
            stateFavorite()
        })

        detailMovieViewModel.errorMovieById.observe(this, { errorMovieById ->
            timberE(errorMovieById)
        })

        detailMovieViewModel.resultTVShowById.observe(this, { tvShowById ->
            tvShow = tvShowById
            stateFavorite()
        })

        detailMovieViewModel.errorTVShowById.observe(this, { errorTVShowById ->
            timberE(errorTVShowById)
        })

        detailMovieViewModel.resultInsertMovie.observe(this, {
            showToast(R.string.txt_movie_add)
        })

        detailMovieViewModel.errorInsertMovie.observe(this, { errorInsertMovie ->
            timberE(errorInsertMovie)
        })

        detailMovieViewModel.resultInsertTVShow.observe(this, {
            showToast(R.string.txt_movie_add)
        })

        detailMovieViewModel.errorInsertTVShow.observe(this, { errorInsertTVShow ->
            timberE(errorInsertTVShow)
        })

        detailMovieViewModel.resultDeleteMovie.observe(this, {
            showToast(R.string.txt_movie_remove)
        })

        detailMovieViewModel.errorDeleteMovie.observe(this, { errorDeleteMovie ->
            timberE(errorDeleteMovie)
        })

        detailMovieViewModel.resultDeleteTVShow.observe(this, {
            showToast(R.string.txt_movie_remove)
        })

        detailMovieViewModel.errorDeleteTVShow.observe(this, { errorDeleteTVShow ->
            timberE(errorDeleteTVShow)
        })
    }

    private fun initClick() {
        binding.setClickFavoriteListener {
            clickFavorite()
        }
    }

    private fun clickFavorite() {
        if (isFavorite) {
            if (movieId != 0) {
                removeFavorite(movie)
            } else {
                removeFavorite(null, tvShow)
            }
        } else {
            detailFavorite?.let { detail -> addFavorite(detail) }
        }

        isFavorite = !isFavorite
        setFavorite()
        binding.btnFavorite.playAnimation()
    }

    private fun showDetail(detail: Detail) {
        binding.txtTitleDetail.text = detail.title
        binding.txtRateMovie.text = detail.rating.toString()
        val rating = (detail.rating / 2).toFloat()
        binding.rbRatingDetail.rating = rating

        if (detail.genre.isEmpty()) {
            binding.txtGenreDetail.text = getString(R.string.txt_no_genre)
        } else {
            CommonUtils.getGenresString(detail.genre, binding.txtGenreDetail)
        }

        if (detail.overview.isEmpty() || detail.overview == "") {
            binding.txtDescDetail.text = getString(R.string.txt_no_desc)
        } else {
            binding.txtDescDetail.text = detail.overview
        }

        loadFromUrl(binding.ivImgDetail, detail.image)

        if (detail.imageBackground != "null") {
            loadFromUrl(binding.ivCoverDetail, detail.imageBackground)
        } else {
            loadFromUrl(binding.ivCoverDetail, detail.image)
        }

        binding.txtDateDetail.text = CommonUtils.toDateString(detail.date)
    }

    private fun initInjector() {
        DaggerDetailComponent.builder()
            .appComponent((application as MovieCatalogueApp).appComponent)
            .build()
            .inject(this)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            onBackPressed()
        }
        return super.onOptionsItemSelected(item)
    }

    public override fun onResume() {
        super.onResume()
        checkConnection(this)
        binding.shimmerViewContainer.startShimmer()
    }

    public override fun onPause() {
        binding.shimmerViewContainer.stopShimmer()
        super.onPause()
    }

    private fun addFavorite(detailFavorite: Detail) {
        if (movieId != 0) {
            detailFavorite.let {
                val movieEntity = MoviesDB(
                    it.id,
                    it.title,
                    it.originalTitle,
                    it.image,
                    it.imageBackground,
                    binding.txtGenreDetail.text.toString(),
                    binding.txtRateMovie.text.toString(),
                    it.date,
                    it.overview
                )
                detailMovieViewModel.getInsertFavoriteMovieContract(movieEntity)
            }
        } else {
            detailFavorite.let {
                val tvShowEntity = TVShowsDB(
                    it.id,
                    it.title,
                    it.originalTitle,
                    it.image,
                    it.imageBackground,
                    binding.txtGenreDetail.text.toString(),
                    binding.txtRateMovie.text.toString(),
                    it.date,
                    it.overview
                )
                detailMovieViewModel.getInsertFavoriteTVSHowContract(tvShowEntity)
            }
        }
    }

    private fun removeFavorite(movie: MoviesDB? = null, tvShow: TVShowsDB? = null) {
        if (movieId != 0) {
            movie?.let {
                detailMovieViewModel.getDeleteFavoriteMovieContract(it)
            }
        } else {
            tvShow?.let {
                detailMovieViewModel.getDeleteFavoriteTVShowContract(it)
            }
        }

    }

    private fun stateFavorite() {
        if (movieId != 0) {
            movie?.let {
                isFavorite = true
                setFavorite()
            }
        } else {
            tvShow?.let {
                isFavorite = true
                setFavorite()
            }
        }
    }

    private fun setFavorite() {
        if (isFavorite) {
            binding.btnFavorite.speed = 1f
            binding.btnFavorite.progress = 1f
        } else {
            binding.btnFavorite.progress = 0f
            binding.btnFavorite.speed = -2f
        }
    }

    companion object {
        const val DETAIL_EXTRA_MOVIE = "extra_movie"
        const val DETAIL_EXTRA_TV_SHOW = "extra_tv_show"

    }
}